pkg/kubelet/deviceplugin/manager.go

```diff
@@ -212,21 +213,38 @@ func (m *ManagerImpl) addEndpoint(r *pluginapi.RegisterRequest) {
 		return
 	}
 
-	stream, err := e.list()
-	if err != nil {
-		glog.Errorf("Failed to List devices for plugin %v: %v", r.ResourceName, err)
-		return
-	}
+	closeChan := make(chan bool)
+	go e.run(closeChan)
 
 	go func() {
-		e.listAndWatch(stream)
-
-		m.mutex.Lock()
-		if old, ok := m.Endpoints[r.ResourceName]; ok && old == e {
-			delete(m.Endpoints, r.ResourceName)
+		// TODO: Consider making timeout value configurable.
+		t := time.NewTimer(5 * time.Second)
+		defer t.Stop()
+
+		for {
+			select {
+			case <-t.C:
+				m.mutex.Lock()
```

Comment:
@lichuqiang, could you take a look at my PR https://github.com/kubernetes/kubernetes/pull/52290 that is to fix the race condition during re-registration? A unit test in manager-test.go is added in that PR for this. The problem is also exposed through the PR that extended e2e gpu test to exercise device plugin restarts: https://github.com/kubernetes/kubernetes/pull/52375.

Reply:
Yes, I see your PR, a good catch, to test is always a good way to spot problems:)
but I don't think the problem has relationship with changes here, it's mostly caused by the lack of callback and corresponding update in node capacity when removing endpoint, as you fixed in your PR. And the change @RenaudWasTaken suggested here is just to handle rededundant plugin that doesn't register devices indeed.
I would prefer revert the PR to it's original state just for the combination of func list & listwatch, and maybe open another PR about the registration issue if necessary, thanks.

Reply:
+1 for separating them into different PRs.